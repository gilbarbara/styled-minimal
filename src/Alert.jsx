import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';

import { px, themeGet } from './utils/helpers';
import { baseStyles, sizesAllPropTypes, variantPropTypes } from './utils/system';

import Box from './Box';

const styles = (props: Object): string => {
  const { size } = props;
  const { borderRadius, maxWidth, padding } = themeGet(props, 'alert');

  return css`
    ${baseStyles.variant};
    border-radius: ${px(borderRadius)};
    font-size: ${baseStyles.fontSize};
    line-height: ${baseStyles.lineHeight};
    max-width: ${px(maxWidth)};
    padding: ${px(padding[size][0])} ${px(padding[size][1])};
    width: 100%;

    a {
      ${baseStyles.color};
    }
  `;
};

const Alert = styled(Box).attrs({
  role: 'alert',
})`
  ${styles};
`;

Alert.displayName = 'Alert';

Alert.propTypes = {
  as: PropTypes.string,
  children: PropTypes.node.isRequired,
  dark: PropTypes.bool,
  outline: PropTypes.bool,
  /** button size */
  size: sizesAllPropTypes,
  /** button variant */
  variant: variantPropTypes,
  ...Box.propTypes,
};

Alert.defaultProps = {
  outline: false,
  size: 'md',
  variant: 'primary',
};

export default Alert;
