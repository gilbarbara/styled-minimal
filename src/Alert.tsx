import * as React from 'react';
import styled, { css } from 'styled-components';

import { getTheme, px } from './utils/helpers';
import {
  baseStyles,
  colorStyles,
  lineHeightStyles,
  sx,
  system,
  variantStyles,
} from './utils/system';

import { AlertProps } from './types';

export const StyledAlert = styled('div')<AlertProps>(
  props => {
    const { borderRadius, maxWidth, padding } = getTheme(props, 'alert');

    return css`
      ${baseStyles};
      ${variantStyles};
      border-radius: ${px(borderRadius)};
      font-size: inherit;
      line-height: ${lineHeightStyles};
      max-width: ${px(maxWidth)};
      padding: ${padding};
      width: 100%;

      a {
        ${colorStyles};
      }
    `;
  },
  sx,
  system,
);

export const Alert = React.forwardRef<HTMLDivElement, AlertProps>((props, ref) => (
  <StyledAlert ref={ref} {...props} />
));

Alert.defaultProps = {
  dark: false,
  invert: false,
  role: 'alert',
  variant: 'primary',
};
