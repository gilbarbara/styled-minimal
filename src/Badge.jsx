import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';

import { px, themeGet } from './utils/helpers';
import { baseStyles, sizesAllPropTypes, variantPropTypes } from './utils/system';

import Box from './Box';

export const Badge = styled(Box)`
  ${(props: Object): string => {
    const {
      borderRadius,
      fontSize,
      fontWeight,
      padding,
    } = themeGet(props, 'badge');
    const fontSizeProp = themeGet(props, 'componentSizes', { key: 'size' });

    return css`
      ${baseStyles.variant};
      border-radius: ${px(borderRadius)};
      display: inline-flex;
      font-size: ${fontSize};
      ${fontSizeProp ? `font-size: ${fontSizeProp}` : ''};
      font-weight: ${fontWeight};
      line-height: 1;
      padding: ${px(padding[0])} ${px(padding[1])};
      vertical-align: baseline;
    `;
  }};
`;

Badge.displayName = 'Badge';

Badge.propTypes = {
  as: PropTypes.string,
  children: PropTypes.node.isRequired,
  dark: PropTypes.bool,
  outline: PropTypes.bool,
  size: sizesAllPropTypes,
  variant: variantPropTypes,
  ...Box.propTypes,
};

Badge.defaultProps = {
  as: 'span',
};

export default Badge;
