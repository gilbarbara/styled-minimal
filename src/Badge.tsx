import * as React from 'react';
import styled, { css } from 'styled-components';

import { getTheme, px } from './utils/helpers';
import { baseStyles, sx, system, variantStyles } from './utils/system';

import { BadgeProps } from './types';

export const StyledBadge = styled('sup')<BadgeProps>(
  props => {
    const { dot } = props;
    const { borderRadius, fontSize, fontWeight, lineHeight, padding, size, sizeDot } = getTheme(
      props,
      'badge',
    );

    return css`
      ${baseStyles};
      ${variantStyles};
      border-radius: ${dot ? '100%' : px(borderRadius)};
      display: inline-flex;
      font-size: ${dot ? 0 : px(fontSize)};
      font-weight: ${fontWeight};
      justify-content: center;
      line-height: ${dot ? 0 : px(lineHeight)};
      height: ${px(dot ? sizeDot : size)};
      padding: ${dot ? 0 : `0 ${px(padding)}`};
      min-width: ${px(dot ? sizeDot : size)};
    `;
  },
  sx,
  system,
);

export const Badge = React.forwardRef<HTMLSpanElement, BadgeProps>((props, ref) => (
  <StyledBadge ref={ref} {...props} />
));

Badge.defaultProps = {
  dark: false,
  dot: false,
  invert: false,
  variant: 'primary',
};
