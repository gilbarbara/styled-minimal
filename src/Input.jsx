import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';

import { getTheme, isDefined, px } from './utils/helpers';
import { formPseudo, inputTextTypes, outlines } from './utils/system';

import Box from './Box';

const styles = (props: Object): string => {
  const {
    bg,
    bordered,
    borderColor: bc,
    borderRadius: br,
    borderWidth: bw,
    color: cl,
    fontFamily,
    fontSize: fz,
    height: h,
    lineHeight: lh,
    padding: pd,
    size,
    type,
    valid,
    width,
  } = props;
  const {
    backgroundColor,
    borderColor,
    borderRadius,
    borderWidth,
    color,
    fontSize,
    height,
    inlineMargin,
    lineHeight,
    padding,
    validation,
  } = getTheme(props, 'input');

  let currentBorderColor = bc || borderColor;

  if (valid) {
    currentBorderColor = validation.valid;
  } else if (valid === false) {
    currentBorderColor = validation.invalid;
  }

  return css`
    background-color: ${bg || backgroundColor};
    border: ${bordered ? `${px(bw || borderWidth)} solid ${currentBorderColor}` : 0};
    border-radius: ${px(isDefined(br) ? br : borderRadius)};
    color: ${cl || color};
    display: ${!['checkbox', 'radio'].includes(type) ? 'block' : 'inline-block'};
    font-family: ${fontFamily || 'inherit'};
    font-size: ${px(fz || fontSize[size])};
    ${inputTextTypes.includes(type) ? `height: ${h || px(height[size])}` : ''};
    line-height: ${lh || lineHeight};
    ${['checkbox', 'radio'].includes(type) ? `margin: 0 ${px(inlineMargin)} 0 0` : ''};
    padding: ${inputTextTypes.includes(type) ? px(isDefined(pd) ? pd : padding[size]) : 0};
    ${!['checkbox', 'radio', 'color'].includes(type) ? `width: ${width || '100%'}` : ''};

    &[type='file'] {
      font-size: ${px(fontSize[size] - 1)};
      padding: ${px(padding[size] - 1)};
    }
  `;
};

const Input = styled(Box).attrs(({ id, name }) => ({
  id: id || name,
}))`
  ${styles};
  ${formPseudo};
`;

Input.propTypes = {
  accept: PropTypes.string,
  as: PropTypes.string,
  autoComplete: PropTypes.string,
  bordered: PropTypes.bool,
  checked: PropTypes.bool,
  defaultChecked: PropTypes.bool,
  defaultValue: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  disabled: PropTypes.bool,
  id: PropTypes.string,
  maxLength: PropTypes.number,
  minLength: PropTypes.number,
  name: PropTypes.string,
  pattern: PropTypes.string,
  placeholder: PropTypes.string,
  readOnly: PropTypes.bool,
  required: PropTypes.bool,
  size: PropTypes.oneOf(['sm', 'md', 'lg']),
  tabindex: PropTypes.number,
  type: PropTypes.oneOf([
    'checkbox',
    'color',
    'date',
    'email',
    'file',
    'hidden',
    'number',
    'password',
    'radio',
    'search',
    'tel',
    'text',
  ]),
  valid: PropTypes.bool,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  ...Box.basePropTypes,
  ...outlines.propTypes,
};

Input.defaultProps = {
  as: 'input',
  bordered: true,
  size: 'md',
  type: 'text',
  value: undefined,
};

export default Input;
