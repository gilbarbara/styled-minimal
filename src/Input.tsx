import * as React from 'react';
import styled, { css } from 'styled-components';

import {
  getColor,
  getDarkColor,
  getDimmerColor,
  getTextColor,
  getTheme,
  px,
} from './utils/helpers';
import {
  baseStyles,
  pseudoStyles,
  inputTextTypes,
  sx,
  system,
  variantStyles,
} from './utils/system';

import { CheckboxProps, InputProps, RadioProps } from './types';

interface InnerProps extends CheckboxProps, RadioProps {
  category?: 'checkbox' | 'radio';
}

export const StyledInput = styled('input')<any>(
  props => {
    const { size = 'md', type = 'text', valid } = props;
    const {
      backgroundColor,
      borderColor,
      borderRadius,
      borderWidth,
      color,
      fontSize,
      height,
      lineHeight,
      padding,
      validation,
    } = getTheme(props, 'input');

    let currentBorderColor = borderColor;

    if (valid) {
      currentBorderColor = validation.valid;
    } else if (valid === false) {
      currentBorderColor = validation.invalid;
    }

    return css`
      ${baseStyles};
      background-color: ${backgroundColor};
      border: ${px(borderWidth)} solid ${currentBorderColor};
      border-radius: ${px(borderRadius)};
      color: ${color};
      display: block;
      font-family: inherit;
      font-size: ${px(fontSize[size])};
      ${inputTextTypes.includes(type) ? `height: ${px(height[size])}` : ''};
      line-height: ${lineHeight};
      padding: ${inputTextTypes.includes(type) ? px(padding[size]) : 0};
      ${!['color'].includes(type) ? 'width: 100%;' : ''};

      &[type='file'] {
        font-size: ${px(fontSize[size] - 1)};
        padding: ${px(padding[size] - 1)};
      }
    `;
  },
  pseudoStyles,
  sx,
  system,
);

const StyledCheckboxRadioInput = styled('input')`
  left: 0;
  opacity: 0;
  position: absolute;
  top: 0;
  z-index: -1;
`;

const StyledText = styled('span')<InnerProps>(props => {
  const { size, category = 'checkbox' } = props;
  const { sizes } = getTheme(props, category);

  return css`
    font-size: ${px(sizes[size!])};
  `;
});

const StyledElement = styled('span')<InnerProps>(
  props => {
    const { dark, invert, size, category = 'checkbox' } = props;
    const { backgroundColor, borderColor, borderRadius, sizes } = getTheme(props, category);
    const color = getColor(props);
    const currentSize = sizes[size!];
    let after;
    let accentColor = dark ? getDarkColor(props) : getTextColor(props);

    if (invert && !dark) {
      accentColor = color;
    }

    if (category === 'checkbox') {
      after = css`
        border-bottom: 2px solid ${accentColor};
        border-left: 2px solid ${accentColor};
        height: 6px;
        transform: rotate(-45deg) translateY(-100%);
        width: 9px;
      `;
    } else {
      after = css`
        background-color: ${accentColor};
        border-radius: 50%;
        height: 6px;
        transform: translate(-50%, -50%);
        width: 6px;
      `;
    }

    return css`
      ${baseStyles};
      border-radius: ${px(borderRadius)};
      ${variantStyles};
      display: inline-flex;
      height: ${px(currentSize)};
      margin-right: 6px;
      position: relative;
      transition: background-color 0.2s, border-color 0.2s;
      width: ${px(currentSize)};

      &:after {
        content: '';
        display: none;
        position: absolute;
        left: 50%;
        top: 50%;
        ${after};
      }

      input:checked ~ & {
        &:after {
          display: inline-block;
        }
      }

      input:disabled ~ & {
        opacity: 0.6;
      }

      input:not(:checked) ~ & {
        background-color: ${backgroundColor};
        border-color: ${borderColor};
      }

      input:not(:checked):disabled ~ & {
        background-color: ${borderColor};
      }
    `;
  },
  sx,
  system,
);

const StyledLabel = styled('label')<InnerProps>(props => {
  const { disabled, inline, category = 'checkbox' } = props;
  const color = getDimmerColor(getColor(props));
  const { borderColor } = getTheme(props, category);

  return css`
    align-items: center;
    cursor: ${disabled ? 'default' : 'pointer'};
    display: ${inline ? 'inline-flex' : 'flex'};
    margin-bottom: 8px;
    position: relative;
    margin-right: ${inline && '12px'};

    &:focus,
    &:hover {
      ${StyledCheckboxRadioInput} {
        &:not(:checked) ~ ${StyledElement} {
          background-color: ${borderColor};
        }

        &:not(:disabled):checked ~ ${StyledElement} {
          background-color: ${color};
          border-color: ${color};
        }
      }
    }
  `;
});

export const Checkbox = React.forwardRef<HTMLInputElement, CheckboxProps>((props, ref) => {
  const {
    defaultChecked,
    checked,
    children,
    disabled,
    inline,
    invert,
    label,
    onChange,
    size,
    theme,
    value,
    variant,
    ...rest
  } = props;

  return (
    <StyledLabel disabled={disabled} inline={inline} theme={theme} variant={variant}>
      <StyledCheckboxRadioInput
        checked={checked}
        aria-checked={checked || defaultChecked}
        defaultChecked={defaultChecked}
        disabled={disabled}
        onChange={onChange}
        ref={ref}
        role="checkbox"
        type="checkbox"
        value={value}
      />
      <StyledElement
        category="checkbox"
        invert={invert}
        size={size}
        theme={theme}
        variant={variant}
        {...rest}
      />
      <StyledText category="checkbox" size={size}>
        {children || label}
      </StyledText>
    </StyledLabel>
  );
});

Checkbox.defaultProps = {
  dark: false,
  inline: false,
  invert: false,
  size: 'md',
  variant: 'primary',
};

/**
 * Use the RadioGroup component instead of this.
 * RadioGroup accepts an `options` prop that render this component in a group and is responsible for managing state and interactions.
 */
export const Radio = React.forwardRef<HTMLInputElement, RadioProps>(
  (
    {
      defaultChecked,
      checked,
      children,
      disabled,
      id,
      inline,
      invert,
      label,
      name,
      onChange,
      size,
      theme,
      value,
      variant,
      ...rest
    },
    ref,
  ) => (
    <StyledLabel
      disabled={disabled}
      inline={inline}
      theme={theme}
      category="radio"
      variant={variant}
    >
      <StyledCheckboxRadioInput
        checked={checked}
        defaultChecked={defaultChecked}
        disabled={disabled}
        id={id}
        name={name}
        onChange={onChange}
        ref={ref}
        type="radio"
        value={value}
      />
      <StyledElement
        invert={invert}
        role="radio"
        aria-checked={!!(checked || defaultChecked)}
        size={size}
        theme={theme}
        variant={variant}
        {...rest}
        category="radio"
      />
      <StyledText size={size} category="radio">
        {children || label}
      </StyledText>
    </StyledLabel>
  ),
);

Radio.defaultProps = {
  dark: false,
  inline: false,
  invert: false,
  size: 'md',
  variant: 'primary',
};

export const Input = React.forwardRef<HTMLInputElement, InputProps>((props, ref) => (
  <StyledInput ref={ref} {...props} />
));

Input.defaultProps = {
  size: 'md',
  type: 'text',
};
