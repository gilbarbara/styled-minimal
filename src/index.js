export { default as Alert } from './Alert';
export { default as Badge } from './Badge';
export { default as Box } from './Box';
export { default as Button } from './Button';
export { default as Code } from './Code';
export { default as Container } from './Container';
export { default as Embed } from './Embed';
export { default as Flex } from './Flex';
export { default as Fieldset } from './Fieldset';
export { default as Form } from './Form';
export { default as FormGroup } from './FormGroup';
export { default as Group } from './Group';
export { default as Heading } from './Heading';
export { default as Image } from './Image';
export { default as Input } from './Input';
export { default as Label } from './Label';
export { default as Legend } from './Legend';
export { default as Link } from './Link';
export { default as List } from './List';
export { default as Paragraph } from './Paragraph';
export { default as Screen } from './Screen';
export { default as Select } from './Select';
export { default as Switch } from './Switch';
export { default as Table } from './Table';
export { default as Text } from './Text';
export { default as Textarea } from './Textarea';
