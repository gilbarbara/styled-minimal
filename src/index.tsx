export { Alert } from './Alert';
export { Badge } from './Badge';
export { Box } from './Box';
export { Button } from './Button';
export { ButtonGroup } from './ButtonGroup';
export { Code } from './Code';
export { CodeBlock } from './CodeBlock';
export { Container } from './Container';
export { Embed } from './Embed';
export { Flex } from './Flex';
export { Fieldset } from './Fieldset';
export { Form } from './Form';
export { FormGroup } from './FormGroup';
export { Grid } from './Grid';
export { Group } from './Group';
export { Heading } from './Heading';
export { Image } from './Image';
export { Checkbox, Input, Radio } from './Input';
export { Label } from './Label';
export { Legend } from './Legend';
export { Link } from './Link';
export { DL, OL, UL } from './List';
export { Loader } from './Loader';
export { Paragraph } from './Paragraph';
export { RadioGroup } from './RadioGroup';
export { Select } from './Select';
export { Switch } from './Switch';
export { Table } from './Table';
export { Text } from './Text';
export { Textarea } from './Textarea';

export { default as system } from '@styled-system/css';
export { mergeTheme as getTheme, px, responsive } from './utils/helpers';
