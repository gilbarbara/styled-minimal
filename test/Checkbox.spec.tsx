import * as React from 'react';
import { render } from '@testing-library/react';

import { Checkbox } from '../src/Input';

describe('Checkbox', () => {
  it('should render properly', () => {
    const { container } = render(
      <div>
        <Checkbox name="test" label="Checkbox" value="test" defaultChecked />
        <Checkbox name="test2" label="Checkbox 2" value="test2" invert />
      </div>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
