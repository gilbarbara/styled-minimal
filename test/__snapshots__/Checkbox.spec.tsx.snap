// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox should render properly 1`] = `
.c2 {
  left: 0;
  opacity: 0;
  position: absolute;
  top: 0;
  z-index: -1;
}

.c5 {
  font-size: 16px;
}

.c4 {
  box-sizing: border-box;
  border-radius: 4px;
  background-color: #2E2E2E;
  border: 1px solid #2E2E2E;
  color: #fff;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  height: 16px;
  margin-right: 6px;
  position: relative;
  -webkit-transition: background-color 0.2s,border-color 0.2s;
  transition: background-color 0.2s,border-color 0.2s;
  width: 16px;
}

.c4:after {
  content: '';
  display: none;
  position: absolute;
  left: 50%;
  top: 50%;
  border-bottom: 2px solid #fff;
  border-left: 2px solid #fff;
  height: 6px;
  -webkit-transform: rotate(-45deg) translateY(-100%);
  -ms-transform: rotate(-45deg) translateY(-100%);
  transform: rotate(-45deg) translateY(-100%);
  width: 9px;
}

input:checked ~ .c4:after {
  display: inline-block;
}

input:disabled ~ .c4 {
  opacity: 0.6;
}

input:not(:checked) ~ .c4 {
  background-color: #f7f8f9;
  border-color: #ccc;
}

input:not(:checked):disabled ~ .c4 {
  background-color: #ccc;
}

.c6 {
  box-sizing: border-box;
  border-radius: 4px;
  background-color: transparent;
  border: 1px solid #2E2E2E;
  color: #2E2E2E;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  height: 16px;
  margin-right: 6px;
  position: relative;
  -webkit-transition: background-color 0.2s,border-color 0.2s;
  transition: background-color 0.2s,border-color 0.2s;
  width: 16px;
}

.c6:after {
  content: '';
  display: none;
  position: absolute;
  left: 50%;
  top: 50%;
  border-bottom: 2px solid #2E2E2E;
  border-left: 2px solid #2E2E2E;
  height: 6px;
  -webkit-transform: rotate(-45deg) translateY(-100%);
  -ms-transform: rotate(-45deg) translateY(-100%);
  transform: rotate(-45deg) translateY(-100%);
  width: 9px;
}

input:checked ~ .c6:after {
  display: inline-block;
}

input:disabled ~ .c6 {
  opacity: 0.6;
}

input:not(:checked) ~ .c6 {
  background-color: #f7f8f9;
  border-color: #ccc;
}

input:not(:checked):disabled ~ .c6 {
  background-color: #ccc;
}

.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-bottom: 8px;
  position: relative;
}

.c0:focus .c1:not(:checked) ~ .c3,
.c0:hover .c1:not(:checked) ~ .c3 {
  background-color: #ccc;
}

.c0:focus .c1:not(:disabled):checked ~ .c3,
.c0:hover .c1:not(:disabled):checked ~ .c3 {
  background-color: #949494;
  border-color: #949494;
}

<div>
  <label
    class="c0"
  >
    <input
      aria-checked="true"
      checked=""
      class="c1 c2"
      role="checkbox"
      type="checkbox"
      value="test"
    />
    <span
      class="c3 c4"
      name="test"
    />
    <span
      class="c5"
    >
      Checkbox
    </span>
  </label>
  <label
    class="c0"
  >
    <input
      class="c1 c2"
      role="checkbox"
      type="checkbox"
      value="test2"
    />
    <span
      class="c3 c6"
      name="test2"
    />
    <span
      class="c5"
    >
      Checkbox 2
    </span>
  </label>
</div>
`;
